File: ios/QuickPic/Models/Friend.swift
import Foundation

struct Friend: Codable, Identifiable {
    let userID: UUID
    let username: String
    let publicKey: String
    let since: Date

    var id: UUID { userID }

    enum CodingKeys: String, CodingKey {
        case userID = "user_id"
        case username
        case publicKey = "public_key"
        case since
    }
}

struct FriendRequest: Codable, Identifiable {
    let id: UUID
    let fromUserID: UUID
    let toUserID: UUID
    let status: String
    let createdAt: Date
    let fromUser: User

    enum CodingKeys: String, CodingKey {
        case id
        case fromUserID = "from_user_id"
        case toUserID = "to_user_id"
        case status
        case createdAt = "created_at"
        case fromUser = "from_user"
    }
}

struct SendFriendRequestRequest: Codable {
    let username: String
}

struct FriendRequestActionRequest: Codable {
    let requestID: UUID

    enum CodingKeys: String, CodingKey {
        case requestID = "request_id"
    }
}

File: server/internal/models/friend.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type FriendRequestStatus string

const (
	FriendRequestPending  FriendRequestStatus = "pending"
	FriendRequestAccepted FriendRequestStatus = "accepted"
	FriendRequestRejected FriendRequestStatus = "rejected"
)

type FriendRequest struct {
	ID         uuid.UUID           `json:"id"`
	FromUserID uuid.UUID           `json:"from_user_id"`
	ToUserID   uuid.UUID           `json:"to_user_id"`
	Status     FriendRequestStatus `json:"status"`
	CreatedAt  time.Time           `json:"created_at"`
}

type FriendRequestWithUser struct {
	FriendRequest
	FromUser UserPublic `json:"from_user"`
}

type Friendship struct {
	ID        uuid.UUID `json:"id"`
	UserAID   uuid.UUID `json:"user_a_id"`
	UserBID   uuid.UUID `json:"user_b_id"`
	CreatedAt time.Time `json:"created_at"`
}

type Friend struct {
	UserID    uuid.UUID `json:"user_id"`
	Username  string    `json:"username"`
	PublicKey string    `json:"public_key"`
	Since     time.Time `json:"since"`
}

type SendFriendRequestRequest struct {
	Username string `json:"username" binding:"required"`
}

type FriendRequestActionRequest struct {
	RequestID uuid.UUID `json:"request_id" binding:"required"`
}
